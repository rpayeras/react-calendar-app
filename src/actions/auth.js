import Swal from "sweetalert2";
import { fetchNoToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = (email, password) => {
    return async (dispatch) => {
        try{
            const res = await fetchNoToken('auth', {email, password}, 'POST')
            const {ok, data, msg} = await res.json()

            if(ok){
                const {name, uid, token} = data;

                localStorage.setItem('token', token)
                localStorage.setItem('token-created', new Date().getTime())
                
                dispatch(login({
                    uid,
                    name
                }))
            } else {
                Swal.fire('Error',msg, 'error')
            }
        } catch(err){
            Swal.fire('Error', err, 'error')
        }
    }
}

export const startRegister = (name, email, password) => {
    return async (dispatch) => {
        try{
            const res = await fetchNoToken('auth/new', {name, email, password}, 'POST')
            const {ok, data, msg, errors} = await res.json()
            let errMsg = '';
            
            if(errors){
                errors.forEach(({msg}, idx) => {
                    errMsg += `\n${idx}: ${msg} \n`
                })
            }

            if(ok){
                const {name, uid, token} = data;

                localStorage.setItem('token', token)
                localStorage.setItem('token-created', new Date().getTime())
                
                dispatch(register({
                    uid,
                    name
                }))
            } else {
                Swal.fire('Error', errMsg, 'error')
            }
        } catch(err){
            Swal.fire('Error', err, 'error')
        }
    }
}

export const startChecking = () => {
    return async (dispatch) => {
        try{
            const res = await fetchWithToken('auth/renew')
            const {ok, data, msg} = await res.json()

            if(ok){
                const {name, uid, token} = data;

                localStorage.setItem('token', token)
                localStorage.setItem('token-created', new Date().getTime())
                
                dispatch(login({
                    uid,
                    name
                }))
            } else {
                console.log(msg)
                //Swal.fire('Error', msg, 'error')
            }

            dispatch(checkingFinish())
        } catch(err){
            console.log(err)
            dispatch(checkingFinish())
            // Swal.fire('Error', err, 'error')
        }
    }
}

export const startLogout = () => {
    return (dispatch) => {
        localStorage.clear()
        dispatch(logout())
    }
}

const checkingFinish = () => ({ type: types.authCheckingFinish })

export const login = (user) => ({
    type: types.authLogin,
    payload: user
})

export const register = (user) => ({
    type: types.authRegister,
    payload: user
})

export const logout = () => ({
    type: types.authLogout
})
