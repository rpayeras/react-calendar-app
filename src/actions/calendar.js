import { types } from "../types/types"
import {fetchWithToken} from "../helpers/fetch"
import { prepareEvents } from "../helpers/prepareEvents";
import Swal from "sweetalert2";

export const eventStartLoading = () => {
    return async (dispatch) => {

        try{
            const res = await fetchWithToken('events')
            const {ok, data} = await res.json()


            if(ok){
                let events = prepareEvents(data)
                dispatch(eventLoaded(events))
            }
        } catch(err) {
            console.log(err)
        }
    }
}

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
})

export const eventStartAddNew = (event) => {
    return async(dispatch, getState) => {
        const {uid, name} = getState().auth;

        try{
            const res = await fetchWithToken('events', event, 'POST')
            const {ok, data} = await res.json();

            if(ok){
                event.id = data.id;
                event.user = {
                    _id: uid,
                    name
                }

                dispatch(eventAddNew(event))
            }
        } catch (err) {
            console.log(err);
        }
    }
}

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
})

export const eventSetCurrent = (event) => ({
    type: types.eventSetCurrent,
    payload: event
});

export const eventCleanCurrent = (event) => ({
    type: types.eventCleanCurrent,
});

export const eventStartUpdateCurrent = (event) => {
    return async(dispatch) => {
        try{
            const res = await fetchWithToken(`events/${event.id}`, event, 'PUT')
            const body = await res.json()

            if(body.ok){
                dispatch(eventUpdateCurrent(event))
            } else {
                Swal.fire('Error', body.msg, 'error')
            }
        } catch(err){
            console.log(err)
        }
    }
}

export const eventUpdateCurrent = (event) => ({
    type: types.eventUpdateCurrent,
    payload: event
});

export const eventStartDelete = () => {
    return async(dispatch, getState) => {
        const {id} = getState().calendar.current
        
        try{
            const res = await fetchWithToken(`events/${id}`, {}, 'DELETE')
            const body = await res.json()

            if(body.ok){
                dispatch(eventDelete())
            } else {
                Swal.fire('Error', body.msg, 'error')
            }
        } catch(err){
            console.log(err)
        }
    }
}

export const eventDelete = () => ({
    type: types.eventDelete
})

export const eventLogout = () => ({
    type: types.eventLogout
})