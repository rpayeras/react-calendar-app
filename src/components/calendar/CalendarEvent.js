import React from "react";
import PropTypes from "prop-types";

export const CalendarEvent = ({ event }) => {
  const { title, user } = event;

  return (
    <div>
      <strong>{title}</strong>
      <span>{user.name}</span>
    </div>
  );
};

CalendarEvent.propTypes = {
  event: PropTypes.object,
};
