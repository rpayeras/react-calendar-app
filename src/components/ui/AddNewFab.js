import { PlusIcon } from "@heroicons/react/solid";
import React from "react";
import { useDispatch } from "react-redux";
import { uiOpenModal } from "../../actions/ui";

export const AddNewFab = () => {
  const dispatch = useDispatch();

  const handleClick = () => {
    dispatch(uiOpenModal());
  };

  return (
    <button
      className="absolute right-2 bottom-2 w-12 h-12 p-2 bg-blue-500 text-white rounded-full shadow-lg hover:bg-white hover:text-blue-500 transition-colors"
      onClick={handleClick}
    >
      <PlusIcon className="" />
    </button>
  );
};
