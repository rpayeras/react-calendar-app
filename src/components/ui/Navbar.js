import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startLogout } from "../../actions/auth";
import { eventLogout } from "../../actions/calendar";
import { LogoutIcon } from "@heroicons/react/solid";

export const Navbar = () => {
  const { name } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const handleLogout = () => {
    dispatch(eventLogout());
    dispatch(startLogout());
  };

  return (
    <div className="shadow-md px-2 py-2 flex justify-between items-center mb-4">
      <span className="text-slate-500 font-bold">Welcome {name}</span>

      <button
        className="flex items-center border border-red-500 text-red-500 text-bold bg-transparent px-2 py-1 rounded-md hover:bg-red-500 hover:text-white transition-colors"
        onClick={handleLogout}
      >
        <LogoutIcon className="w-4 h-4" />
        Exit
      </button>
    </div>
  );
};
