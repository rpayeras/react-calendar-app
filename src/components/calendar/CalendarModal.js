import React, { useEffect, useState } from 'react'
import Modal from 'react-modal';
import DateTimePicker from 'react-datetime-picker';
import moment from 'moment'
import { useForm } from '../../hooks/useForm';
import Swal from 'sweetalert2';
import { useDispatch, useSelector } from 'react-redux';
import { uiCloseModal } from '../../actions/ui';
import { eventCleanCurrent, eventStartAddNew, eventStartUpdateCurrent } from '../../actions/calendar';

const customStyles = {
    content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
    },
};

if(process.env.NODE_ENV !== 'test') {
    Modal.setAppElement('#root');
}

const initEvent = {
    title: '',
    startDate: moment().minutes(0).seconds(0).add(1, 'hours').toDate(),
    endDate: moment().minutes(0).seconds(0).add(2, 'hours').toDate(),
    bgcolor: '#FAFAFA',
    notes: '',
    user: {
        id: '123qwe',
        name: 'User'
    }
}

export const CalendarModal = () => {
    const {current} = useSelector(state => state.calendar)
    
    const [formValues, setFormValue, resetForm] = useForm(initEvent)

    const {modalOpen} = useSelector(state => state.ui)
    const [titleValid, setTitleValid] = useState(true)
    const {id, startDate, endDate, title, notes} = formValues;

    const dispatch = useDispatch();

    useEffect(() => {
        if(current){
            resetForm(current)
        } else {
            resetForm()
        }
    }, [current])
    
    const handleClose = () => {
        dispatch(uiCloseModal())
        dispatch(eventCleanCurrent())
    }

    const handleStartDate = (e) => {
        setFormValue({
            target: {
                name: 'startDate',
                value: e
            }
        })
    }

    const handleEndDate = (e) => {
        setFormValue({
            target: {
                name: 'endDate',
                value: e
            }
        })
    }

    const handleSubmit = (e) => {
        e.preventDefault()

        const momentStart = moment(startDate)
        const momentEnd = moment(endDate)

        if(momentStart.isSameOrAfter(momentEnd)){
            Swal.fire('Error', 'Date end should be major than start')
        }

        if(title.trim().length < 2) {
            return setTitleValid(false)
        } else {
            setTitleValid(true)
        }

        //If id not exists we gonna create a new event
        if(!id){
            dispatch(eventStartAddNew(formValues))
        } else {
            dispatch(eventStartUpdateCurrent({
                ...formValues,
            }))
        }

        dispatch(uiCloseModal())
        dispatch(eventCleanCurrent())
        resetForm()
    }

    return (
        <Modal
            isOpen={modalOpen}
            // onAfterOpen={afterOpenModal}
            onRequestClose={handleClose}
            style={customStyles}
            closeTimeoutMS={200}
            className="modal"
            overlayClassName="modal-fondo"
            ariaHideApp={!process.env.NODE_ENV == 'test'}
        >
        <h1>{current ? 'Updating' : 'New'} Event</h1>
        <hr />
        <form className="container" onSubmit={handleSubmit}>

            <div className="form-group">
                <label>Date and hour of start</label><br />
                <DateTimePicker 
                    onChange={handleStartDate} 
                    value={startDate} 
                    minDate={startDate}
                    />
            </div>

            <div className="form-group">
                <label>Date and hour of finish</label><br />
                <DateTimePicker 
                    onChange={handleEndDate} 
                    value={endDate} 
                    minDate={startDate}
                    />
            </div>

            <hr />
            <div className="form-group">
                <label>Title and notes</label>
                <input 
                    type="text" 
                    className={`form-control ${!titleValid ? 'is-invalid' : ''}`}
                    placeholder="Title of event"
                    name="title"
                    autoComplete="off"
                    value={title}
                    onChange={setFormValue}
                />
                <small id="emailHelp" className="form-text text-muted">Short description</small>
            </div>

            <div className="form-group">
                <textarea 
                    type="text" 
                    className="form-control"
                    placeholder="Notes"
                    rows="5"
                    name="notes"
                    value={notes}
                    onChange={setFormValue}
                ></textarea>
                <small id="emailHelp" className="form-text text-muted">Aditional info</small>
            </div>

            <button
                type="submit"
                className="btn btn-outline-primary btn-block"
            >
                <i className="far fa-save"></i>
                <span>Save</span>
            </button>
        </form>
    </Modal>
    )
}
