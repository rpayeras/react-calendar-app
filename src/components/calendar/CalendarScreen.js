import React, { useEffect, useState } from "react";
import "react-big-calendar/lib/css/react-big-calendar.css";
import moment from "moment";
import "moment/locale/es";

import { Calendar, momentLocalizer } from "react-big-calendar";
import { messages } from "../../helpers/calendar-messages";

import { CalendarEvent } from "./CalendarEvent";
import { CalendarModal } from "./CalendarModal";
import { Navbar } from "../ui/Navbar";

import { uiOpenModal } from "../../actions/ui";
import { useSelector, useDispatch } from "react-redux";
import {
  eventCleanCurrent,
  eventSetCurrent,
  eventStartLoading,
} from "../../actions/calendar";
import { AddNewFab } from "../ui/AddNewFab";
import { DeleteEventFab } from "../ui/DeleteEventFab";

moment.locale("es");
const localizer = momentLocalizer(moment);

export const CalendarScreen = () => {
  const [lastView, setLastView] = useState(
    localStorage.getItem("lastView") || "month"
  );
  const dispatch = useDispatch();
  const { events, current } = useSelector((state) => state.calendar);
  const { uid } = useSelector((state) => state.auth);

  useEffect(() => {
    dispatch(eventStartLoading());
  }, [dispatch]);

  const onDoubleClick = (e) => {
    dispatch(uiOpenModal());
  };

  const onViewChange = (e) => {
    localStorage.setItem("lastView", e);
    setLastView(e);
  };

  const onSelectEvent = (e) => {
    dispatch(eventSetCurrent(e));
  };

  const onSelectedSlot = (e) => {
    dispatch(eventCleanCurrent());
    dispatch(uiOpenModal());
  };

  const eventStyleGetter = (event, start, end, isSelected) => {
    const style = {
      backgroundColor: uid === event.user._id ? "#367CF7" : "#464646",
      borderRadius: "0px",
      opacity: 0.8,
      display: "block",
      color: "white",
    };

    return {
      style,
    };
  };

  return (
    <section className="calendar-screen">
      <Navbar />
      <Calendar
        localizer={localizer}
        events={events}
        startAccessor="startDate"
        endAccessor="endDate"
        style={{ height: 500 }}
        messages={messages}
        eventPropGetter={eventStyleGetter}
        onDoubleClickEvent={onDoubleClick}
        onSelectEvent={onSelectEvent}
        onView={onViewChange}
        onSelectedSlot={onSelectedSlot}
        selectable={true}
        components={{
          event: CalendarEvent,
        }}
        view={lastView}
      />
      <AddNewFab />
      {current && <DeleteEventFab />}

      <CalendarModal />
    </section>
  );
};
