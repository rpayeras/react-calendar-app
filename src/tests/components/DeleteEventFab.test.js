import React from 'react'
import { mount } from "enzyme"
import { Provider } from "react-redux"

import configureStore from "redux-mock-store"
import thunk from "redux-thunk"

import { DeleteEventFab } from '../../components/ui/DeleteEventFab'
import { eventStartDelete } from '../../actions/calendar'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {}

let store = mockStore(initState);
store.dispatch = jest.fn()

jest.mock('../../actions/calendar', () => ({
    eventStartDelete: jest.fn()
}));

const wrapper = mount(
    <Provider store={store}>
        <DeleteEventFab />
    </Provider>
)

describe('Testing DeleteEventFab component', () => { 

    beforeEach(() => {
        store = mockStore(initState)
        jest.clearAllMocks()
    })

    test('should render', () => { 
        expect(wrapper).toMatchSnapshot()
     })

     test('should dispatch eventStartDelete', () => {
         wrapper.find('button').prop('onClick')()

         expect(eventStartDelete).toHaveBeenCalled()
     })
 })