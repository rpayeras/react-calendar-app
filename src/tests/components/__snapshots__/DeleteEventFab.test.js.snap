// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing DeleteEventFab component should render 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
    Delete
  </span>
</button>
`;
